src/codegen/abi.js

```diff
@@ -411,37 +430,93 @@ module.exports = class AbiCodeGenerator {
           .forEach(member => returnType.addMember(member))
 
+        //Create types for Tuple outputs
+        outputs.forEach((output, index) => {
+          if (output.get('type') === 'tuple') {
+            types = types.push(
+              this._generateTupleType(
+                output,
+                index,
+                tupleResultParentType,
+                'function',
+                this.abi.name,
+              ).classes,
+            )
+          }
+        })
+
         // Add the type to the types we'll create
         types = types.push(returnType)
 
         returnType = tsCodegen.namedType(returnType.name)
       } else {
-        returnType = tsCodegen.namedType(
-          typesCodegen.ascTypeForEthereum(
-            member
-              .get('outputs')
+        let type = outputs
+          .get(0)
+          .get('type')
+        if (type === 'tuple') {
+          // Add the Tuple type to the types we'll create
+          types = types.push(
+            this._generateTupleType(
+              outputs.get(0),
+              0,
+              tupleResultParentType,
+              'function',
+              this.abi.name,
+            ).classes,
+          )
+
+          returnType = this._generateTupleClassName(
+            tupleResultParentType,
+            outputs
               .get(0)
-              .get('type'),
-          ),
-        )
+              .get('name'),
+            0,
+          )
+        } else {
+          returnType = tsCodegen.namedType(typesCodegen.ascTypeForEthereum(type))
+        }
       }
 
       // Disambiguate inputs with duplicate names
       let inputs = util.disambiguateNames({
         values: member.get('inputs', immutable.List()),
         getName: (input, index) => input.get('name') || `param${index}`,
-        setName: (input, name) => input.set('_alias', name),
+        setName: (input, name) => input.set('name', name),
+      })
+
+      // Generate a type suffix to identify the Tuple inputs to a function
+      let tupleInputParentType = this.abi.name + '__' + fnAlias + 'Input__'
+
+      //Create types for Tuple inputs
+      inputs.forEach((input, index) => {
+        if (input.get('type') === 'tuple') {
+          types = types.push(
+            this._generateTupleType(
+              input,
+              index,
+              tupleInputParentType,
+              'function',
+              this.abi.name,
+            ).classes,
+          )
+        }
       })
 
       // Generate and add a method that implements calling the function on
       // the smart contract
       klass.addMethod(
         tsCodegen.method(
           fnAlias,
-          inputs.map(input =>
+          inputs.map((input, index) =>
             tsCodegen.param(
-              input.get('_alias'),
-              typesCodegen.ascTypeForEthereum(input.get('type')),
+              input.get('name'),
+              input.get('type') === 'tuple'
+                ? this._generateTupleClassName(
```

Comment:
Same as before: Is this guaranteed to be compatible with the `_generateTupleType` result? I think we should combine the type + name generation.